droidlet/tools/hitl/nsp_retrain_infra.py

```diff
@@ -0,0 +1,259 @@
+import logging
+import os
+import shutil
+import time
+from subprocess import Popen, PIPE, TimeoutExpired
+import boto3
+from datetime import datetime
+
+from typing import List
+
+from utils.data_generator import DataGenerator
+from utils.job_listener import JobListener
+from utils.task_runner import TaskRunner
+
+
+log_formatter = logging.Formatter(
+    "%(asctime)s [%(filename)s:%(lineno)s - %(funcName)s() %(levelname)s]: %(message)s"
+)
+logger = logging.getLogger()
+logger.handlers.clear()
+logger.setLevel("INFO")
+sh = logging.StreamHandler()
+sh.setFormatter(log_formatter)
+logger.addHandler(sh)
+
+LISTENER_SLEEP_TIME = 10  # Seconds to wait before listener looks for new data
+NSP_RETRAIN_TIMEOUT = 18000  # Wait a max of 5h for the NSP retraining script to finish
+MODEL_OUTPUT_POLL_TIME = 60  # Seconds to wait between looking for model output logs
+NUM_MODEL_OUTPUT_POLLS = 500  # Number of times to poll model output files before assuming finished
+
+s3 = boto3.client('s3')
+
+NUM_TRAINING_RUNS = 6
+MODEL_NAME = "best_model.pth"
+MODEL_INFO_NAME = "best_model_info.txt"
+
+
+class NSPRetrainingJob(DataGenerator):
+    def __init__(self):
+        super(NSPRetrainingJob, self).__init__()
+        self.batch_id = ""
+
+    # Functions originally from sweep_monitor.py:
+    def max_and_argmax(self,l):
+        i = max(range(len(l)), key=lambda i: l[i])
+        return l[i], i
+
+    def get_accs(self):
+        accs = {}
+        logs = [l for l in os.listdir() if l[-4:] == ".log"]
+        if len(logs) < NUM_TRAINING_RUNS:  # Wait until all logs have appeared, takes ~30 min from first to last
+            return False
+        for log in logs:
+            accs[log] = []
+            f = open(log)
+            l = f.readlines()
+            for line in l:
+                if "evaluating on" in line:
+                    s = line.split("\t")
+                    a = s[2][11:17]
+                    accs[log].append(float(a))
+        return accs
+
+    def copy_best_model(self,accs):
+        macc = 0.0
+        mname = ""
+        midx = 0
+        for name, acc in accs.items():
+            m, mi = self.max_and_argmax(acc)
+            if m > macc:
+                macc = m
+                mname = name
+                midx = mi
+        source_model = mname[:-4] + "_ep" + str(midx) + ".pth"
+        # TODO some sort of lock/semaphore?
+        shutil.copyfile(source_model, MODEL_NAME)
+        f = open(MODEL_INFO_NAME, "w")
+        f.write(source_model + "\n")
+        f.write("valid acc " + str(macc)  + "\n")
+        f.close()
+
+    def downloadData(self):
+        base_data_dir = os.path.join(opts.droidlet_dir, opts.full_data_dir)
+        batch_id = str(self.batch_id)
+        download_dir = os.path.join(base_data_dir, batch_id) # Currently downloads data to a new dir each time
+        try:
+            os.mkdir(download_dir)
+        except FileExistsError:
+            pass
+        data_filepath = download_dir + '/nsp_data.txt'
```

Comment:
we will only use the new data from this batch to train the model. Remember that `nsp_data.txt` on S3 contains all the data from each HiTL run and we have a meta file for each run which stores the indexes of the data entries of this batch in `nsp_data.txt` file (checkout the notebook quip!). You can use this info to construct the mask required in the model training script.